File: solutions/hash_table/1-4.py
"""
Problem Number 1.4
Problem : Given a string, write an alogorithm to check if it is a permutation of palindrome. A permutation is a rearrangement of letters. The palindromedoes not need to be a dictionary words.
"""

import array

class palindrome(object):

	def check_for_permutation_palindrome(self):
		
		count_char = [0]*128  # Number of ASCII char is 128 

		print ("Please enter the string : ")
		s = input()

		for i in s:
			count_char [ ord(i) ] = count_char [ ord(i) ] + 1

		is_odd = False

		for i in range (0,127):
			
			if count_char[i] % 2 == 1:
				if is_odd == True:
					print ("Permutations of this string can't be a palindrome")
					return
				is_odd = True
		print ("The string's permutation is a palindrome")

p= palindrome()
p.check_for_permutation_palindrome()

File: solutions/hash_table/1-4.cpp
/*
Problem Number 1.4
Problem : Given a string, write an alogorithm to check if it is a permutation of palindrome.A permutation is a rearrangement of letters.The palindromedoes not need to be a dictionary words.

Logic : Count the occurrence of the chars of the string.The count should be even other than one char(uptoo 1 char can be odd)
*/

#include <iostream>
#include <string>

using namespace std;

class palindrome
{
	public:
	void checkPermutationPalindrome()
	{
		cout << "\nPlease enter the string : ";
		string s;
		cin >> s;
		int countChar[128] = { 0 };
		for (int i = 0; i < s.size(); i++)
		{
			countChar[int(s[i])] = countChar[int(s[i])] + 1;
		}

		bool is_odd = false; 

		for (int i = 0; i < 128; i++)
		{
			if (countChar[i] % 2 == 1)
			{
				if (is_odd == true)
				{
					cout << "\nPermutations of this string can't be a palindrome";
					return;
				}
				is_odd = true;
			}
		}
		cout << "\nThe string's permutation is a palindrome";
	}
};

int main()
{
	palindrome p;
	p.checkPermutationPalindrome();
}
